extensions/arc/src/ui/dialogs/connectControllerDialog.ts

```diff
@@ -5,101 +5,144 @@
 
 import { ControllerInfo } from 'arc';
 import * as azdata from 'azdata';
+import * as azdataExt from 'azdata-ext';
 import { v4 as uuid } from 'uuid';
 import * as vscode from 'vscode';
 import { Deferred } from '../../common/promise';
 import * as loc from '../../localizedConstants';
 import { ControllerModel } from '../../models/controllerModel';
 import { InitializingComponent } from '../components/initializingComponent';
 import { AzureArcTreeDataProvider } from '../tree/azureArcTreeDataProvider';
+import { getErrorMessage } from '../../common/utils';
 
 export type ConnectToControllerDialogModel = { controllerModel: ControllerModel, password: string };
-export class ConnectToControllerDialog extends InitializingComponent {
-	private modelBuilder!: azdata.ModelBuilder;
+abstract class ControllerDialogBase extends InitializingComponent {
+	protected modelBuilder!: azdata.ModelBuilder;
+	protected dialog: azdata.window.Dialog;
 
-	private urlInputBox!: azdata.InputBoxComponent;
-	private nameInputBox!: azdata.InputBoxComponent;
-	private usernameInputBox!: azdata.InputBoxComponent;
-	private passwordInputBox!: azdata.InputBoxComponent;
-	private rememberPwCheckBox!: azdata.CheckBoxComponent;
+	protected urlInputBox!: azdata.InputBoxComponent;
+	protected nameInputBox!: azdata.InputBoxComponent;
+	protected usernameInputBox!: azdata.InputBoxComponent;
+	protected passwordInputBox!: azdata.InputBoxComponent;
 
-	private _completionPromise = new Deferred<ConnectToControllerDialogModel | undefined>();
+	protected getComponents(): (azdata.FormComponent<azdata.Component> & { layout?: azdata.FormItemLayout | undefined; })[] {
```

Comment:
I've left this comment in a few other places - but generally if a function is named "getX" that doesn't take any parameters then my immediate thought it "why isn't it just a property getter". Especially when it is just directly returning a value and not really doing any logic.